examples/flava/native/train.py

```diff
@@ -0,0 +1,385 @@
+                # - memory reduction from checkpointing is less than expected
+
+                check_fn = lambda submodule: isinstance(
+                    submodule, TransformerEncoderLayer
+                )
+                non_reentrant_wrapper = partial(
+                    checkpoint_wrapper,
+                    offload_to_cpu=False,
+                    checkpoint_impl=CheckpointImpl.NON_REENTRANT,
+                )
+                apply_activation_checkpointing_wrapper(
+                    model,
+                    checkpoint_wrapper_fn=non_reentrant_wrapper,
+                    check_fn=check_fn,
+                )
+
+                check_fn = lambda submodule: isinstance(
+                    submodule, TransformerEncoderLayer
+                )
+                non_reentrant_wrapper = partial(
+                    checkpoint_wrapper,
+                    offload_to_cpu=False,
+                    checkpoint_impl=CheckpointImpl.REENTRANT,
+                )
+                apply_activation_checkpointing_wrapper(
+                    model,
+                    checkpoint_wrapper_fn=non_reentrant_wrapper,
+                    check_fn=check_fn,
+                )
+            print0(f"after FSDP {torch.cuda.memory_allocated()/1024**3:.3} GB")
+
+        else:
+            raise ValueError(f"unknown strategy: {strategy}")
+
+        print0(
+            f"after {strategy} model parameters: {get_model_parameters(model):,}, "
+            f"size: {get_model_size_gb(model):.3} GB"
+        )
+
+        return model
+
+    def calculate_loss(
+        self, output: FLAVAPretrainingLossOutput, validation=False
+    ) -> torch.Tensor:
+        losses = output.losses
+
+        total_loss = 0
+        for key in losses:
+            if losses[key] is not None:
+                total_loss += losses[key]
+                loss_reduce = losses[key].detach()
+                dist.reduce(loss_reduce, dst=0)
+                if validation:
+                    mode = "validation"
+                else:
+                    mode = "train"
+                self.log(
+                    f"{mode}/losses/{key}",
+                    loss_reduce.item() / dist.get_world_size(),
+                )
+
+        return total_loss
+
+    def preprocess_data(self, data: Dict[str, Any]):
+        data = self.datamodule.on_before_batch_transfer(data, None)
+        data = move_to_device(data, self.device)
+        return self.datamodule.on_after_batch_transfer(data, None)
+
+    def train(self) -> None:
+        print0(OmegaConf.to_container(self.config.training))
+        self.model = self.create_model()
+        model = self.model
+
+        optimizer, scheduler = get_optimizer(
+            model,
+            **self.config.training.optimizer,
+        )
+
+        while True:
+            t0 = time.time()
+            self.epochs += 1
+            dataloader = self.datamodule.train_dataloader()
+            dataloader.set_epoch(self.epochs)
```

Comment:
wat does this do?

Reply:
sets the RNG seed given an epoch for distributedsampler, see the warning here: https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler